pkg/controller/controllercmd/builder.go

```diff
@@ -207,7 +207,7 @@ func (b *ControllerBuilder) Run(ctx context.Context, config *unstructured.Unstru
 		if err := server.PrepareRun().Run(ctx.Done()); err != nil {
 			klog.Error(err)
 		}
-		klog.Fatal("server exited")
```

Comment:
/hold

This won't do graceful shutdown. You can't do graceful termination unless you know that all the business logic is shutdown already. That's why this intentionally calls os.Exit. Graceful termination means having to track all the gorutines running business logic and waiting for them to end, releasing the lock only after they finish. We have been investing some time to that with @mfojtik but it is a long run that needs several fixes in how we start controller and how we need to track gorutines.

Although I don't think you need graceful termination to fix the BZ. ReleaseOnCancel immediately followed by os.Exit / glog.Fatal should release the lock. I think @mfojtik added the ReleaseOnCancel recently - did you check if it is still not releasing it?